nntrainer/layers/fc_layer.cpp

```diff
@@ -33,103 +33,94 @@ namespace nntrainer {
+    bias_dim, bias_initializer, WeightRegularizer::NONE, 1.0f, "FC:bias", true);
+
+  // if (weights.empty()) {
+  //   weights.reserve(2);
+  //   weights.emplace_back(weight_dim, weight_initializer, weight_regularizer,
+  //                        weight_regularizer_constant, true, false,
+  //                        "FC:weight");
+  //   weights.emplace_back(bias_dim, bias_initializer, WeightRegularizer::NONE,
+  //                        1.0f, true, false, "FC:bias");
+  //   manager.trackWeights(weights);
+  // } else {
+  //   weights[FCParams::weight].reset(weight_dim, weight_initializer,
+  //                                   weight_regularizer,
+  //                                   weight_regularizer_constant, true);
+  //   weights[FCParams::bias].reset(bias_dim, bias_initializer,
+  //                                 WeightRegularizer::NONE, 1.0f, true);
+  // }
 }
 
-void FullyConnectedLayer::export_to(Exporter &exporter,
-                                    ExportMethods method) const {
-  LayerV1::export_to(exporter, method);
+void FullyConnectedLayer::exportTo(Exporter &exporter,
+                                   const ExportMethods &method) const {
+  Layer::exportTo(exporter, method);
   exporter.saveResult(fc_props, method, this);
 }
 
-void FullyConnectedLayer::setProperty(const PropertyType type,
-                                      const std::string &value) {
-  switch (type) {
-  case PropertyType::unit: {
-    from_string(value, std::get<props::Unit>(fc_props));
-  } break;
-  default:
-    LayerV1::setProperty(type, value);
-    break;
-  }
+void FullyConnectedLayer::setProperty(const std::vector<std::string> &values) {
+  auto remain_props = loadProperties(values, fc_props);
+  LayerImpl::setProperty(remain_props);
 }
 
-void FullyConnectedLayer::forwarding(bool training) {
-  Tensor &weight =
-    weightAt(static_cast<int>(FCParams::weight)).getVariableRef();
-  Tensor &bias = weightAt(static_cast<int>(FCParams::bias)).getVariableRef();
+void FullyConnectedLayer::forwarding(RunLayerContext &context, bool training) {
+  Tensor &weight = context.getWeight(weight_idx[FCParams::weight]);
+  Tensor &bias = context.getWeight(weight_idx[FCParams::bias]);
+
+  Tensor &hidden_ = context.getOutput(0);
+  Tensor &input_ = context.getInput(0);
 
-  Tensor &hidden_ = net_hidden[0]->getVariableRef();
-  Tensor &input_ = net_input[0]->getVariableRef();
   input_.dot(weight, hidden_);
   hidden_.add_i(bias);
-
-  loss = weightAt(static_cast<int>(FCParams::weight)).getRegularizationLoss();
 }
 
-void FullyConnectedLayer::copy(std::shared_ptr<LayerV1> l) {
-  LayerV1::copy(l);
-
-  std::shared_ptr<FullyConnectedLayer> from =
-    std::static_pointer_cast<FullyConnectedLayer>(l);
-
-  std::get<props::Unit>(fc_props) = std::get<props::Unit>(from->fc_props);
-}
+void FullyConnectedLayer::calcDerivative(RunLayerContext &context) {
+  Tensor &weight = context.getWeight(weight_idx[FCParams::weight]);
```

Comment:
Does `runLayerContext.getWeight(FCParams::weight)` returns the first registered weight?
or the weight index is always dynamic inside layer context?

Reply:
getWeight() will return the weight for the given index. This index is returned when requestWeight() is called.

Reply:
So, my concern is extra level of indirection is always required for every layer.

If returned index from `requiestWeight` is always the new index of an array, we can use `FCParams::weight`. instead of `weight_idx[FCParams::weight]`.

Reply:
This is to ensure that the change in declaration order does not affect the result.

Reply:
I was wondering about the semantic. If that's the case it might be better in this case we promote weight_idx to vector and move to layerImpl later :)

Reply:
Even moving to LayerImpl means that layer_context needs to provide a guarantee that it can always provide weight indices from 0 and in incremental fashion order. For this to work, weights and requested tensor will have to maintain two separate lists (which is done currently, but I am thinking the lists will have to updated based on LIFESPAN type).
So, I will update this when the memory optimizations are done.